A couple of test errors:

```
[ERROR] requireThatInvalidTensorTypeSpecThrowsException(com.yahoo.searchdefinition.processing.RankingExpressionWithTensorTestCase)  Time elapsed: 0.025 s  <<< FAILURE!
java.lang.AssertionError: 

Expected: (an instance of java.lang.IllegalArgumentException and exception with message a string containing "For constant tensor 'my_tensor' in rank profile 'my_profile': Illegal tensor type spec: A tensor type spec must be on the form tensor[<valuetype>]?(dimensionidentifier[{}|[length?]*), but was 'tensor(x)'. Dimension 'x' is on the wrong format. Examples: tensor(x[]), tensor<float>(name{}, x[10])")
     but: exception with message a string containing "For constant tensor 'my_tensor' in rank profile 'my_profile': Illegal tensor type spec: A tensor type spec must be on the form tensor[<valuetype>]?(dimensionidentifier[{}|[length?]*), but was 'tensor(x)'. Dimension 'x' is on the wrong format. Examples: tensor(x[]), tensor<float>(name{}, x[10])" message was "For constant tensor 'my_tensor' in rank profile 'my_profile': Illegal tensor type spec: A tensor type spec must be on the form tensor[<valuetype>]?(dimensionidentifier[{}|[length]*), but was 'tensor(x)'. Dimension 'x' is on the wrong format. Examples: tensor(x[3]), tensor<float>(name{}, x[10])"
Stacktrace was: java.lang.IllegalArgumentException: For constant tensor 'my_tensor' in rank profile 'my_profile': Illegal tensor type spec: A tensor type spec must be on the form tensor[<valuetype>]?(dimensionidentifier[{}|[length]*), but was 'tensor(x)'. Dimension 'x' is on the wrong format. Examples: tensor(x[3]), tensor<float>(name{}, x[10])
```

and

```
[ERROR] requireThatIllegalTensorTypeSpecThrowsException(com.yahoo.searchdefinition.processing.TensorFieldTestCase)  Time elapsed: 0.003 s  <<< FAILURE!
java.lang.AssertionError: 

Expected: (an instance of java.lang.IllegalArgumentException and exception with message a string containing "Field type: Illegal tensor type spec: A tensor type spec must be on the form tensor[<valuetype>]?(dimensionidentifier[{}|[length?]*), but was 'tensor(invalid)'. Dimension 'invalid' is on the wrong format. Examples: tensor(x[]), tensor<float>(name{}, x[10])")
     but: exception with message a string containing "Field type: Illegal tensor type spec: A tensor type spec must be on the form tensor[<valuetype>]?(dimensionidentifier[{}|[length?]*), but was 'tensor(invalid)'. Dimension 'invalid' is on the wrong format. Examples: tensor(x[]), tensor<float>(name{}, x[10])" message was "Field type: Illegal tensor type spec: A tensor type spec must be on the form tensor[<valuetype>]?(dimensionidentifier[{}|[length]*), but was 'tensor(invalid)'. Dimension 'invalid' is on the wrong format. Examples: tensor(x[3]), tensor<float>(name{}, x[10])"
Stacktrace was: java.lang.IllegalArgumentException: Field type: Illegal tensor type spec: A tensor type spec must be on the form tensor[<valuetype>]?(dimensionidentifier[{}|[length]*), but was 'tensor(invalid)'. Dimension 'invalid' is on the wrong format. Examples: tensor(x[3]), tensor<float>(name{}, x[10])
```

--- a/config-model/src/test/java/com/yahoo/searchdefinition/processing/RankingExpressionWithTensorTestCase.java
+++ b/config-model/src/test/java/com/yahoo/searchdefinition/processing/RankingExpressionWithTensorTestCase.java
@@ -2,9 +2,10 @@
 package com.yahoo.searchdefinition.processing;
 
 import com.yahoo.searchdefinition.parser.ParseException;
-import org.junit.Rule;
 import org.junit.Test;
-import org.junit.rules.ExpectedException;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
 
 /**
  * @author geirst
@@ -138,23 +139,29 @@ public class RankingExpressionWithTensorTestCase {
         f.assertRankProperty("tensor(x{})", "constant(my_tensor).type", "my_profile");
     }
 
-    @Rule
-    public ExpectedException exception = ExpectedException.none();
-
     @Test
     public void requireThatInvalidTensorTypeSpecThrowsException() throws ParseException {
-        exception.expect(IllegalArgumentException.class);
-        exception.expectMessage("For constant tensor 'my_tensor' in rank profile 'my_profile': Illegal tensor type spec: A tensor type spec must be on the form tensor[<valuetype>]?(dimensionidentifier[{}|[length?]*), but was 'tensor(x)'. Dimension 'x' is on the wrong format. Examples: tensor(x[]), tensor<float>(name{}, x[10])");
-        RankProfileSearchFixture f = new RankProfileSearchFixture(
-                "  rank-profile my_profile {\n" +
-                "    constants {\n" +
-                "      my_tensor {\n" +
-                "        value: { {x:1}:1 }\n" +
-                "        type: tensor(x)\n" +
-                "      }\n" +
-                "    }\n" +
-                "  }");
-        f.compileRankProfile("my_profile");
+        try {
+            RankProfileSearchFixture f = new RankProfileSearchFixture(
+                    "  rank-profile my_profile {\n" +
+                    "    constants {\n" +
+                    "      my_tensor {\n" +
+                    "        value: { {x:1}:1 }\n" +
+                    "        type: tensor(x)\n" +
+                    "      }\n" +
+                    "    }\n" +
+                    "  }");
+            f.compileRankProfile("my_profile");
+            fail("Expected exception");
+        }
+        catch (IllegalArgumentException e) {
+            assertStartsWith("For constant tensor 'my_tensor' in rank profile 'my_profile': Illegal tensor type spec",
+                             e.getMessage());
+        }
+    }
+
+    private void assertStartsWith(String prefix, String string) {
+        assertEquals(prefix, string.substring(0, Math.min(prefix.length(), string.length())));
     }
 
 }

--- a/config-model/src/test/java/com/yahoo/searchdefinition/processing/TensorFieldTestCase.java
+++ b/config-model/src/test/java/com/yahoo/searchdefinition/processing/TensorFieldTestCase.java
@@ -3,48 +3,68 @@ package com.yahoo.searchdefinition.processing;
 
 import com.yahoo.searchdefinition.SearchBuilder;
 import com.yahoo.searchdefinition.parser.ParseException;
-import org.junit.Rule;
 import org.junit.Test;
-import org.junit.rules.ExpectedException;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
 
 /**
  * @author geirst
  */
 public class TensorFieldTestCase {
 
-    @Rule
-    public ExpectedException exception = ExpectedException.none();
-
     @Test
     public void requireThatTensorFieldCannotBeOfCollectionType() throws ParseException {
-        exception.expect(IllegalArgumentException.class);
-        exception.expectMessage("For search 'test', field 'f1': A field with collection type of tensor is not supported. Use simple type 'tensor' instead.");
-        SearchBuilder.createFromString(getSd("field f1 type array<tensor(x{})> {}"));
+        try {
+            SearchBuilder.createFromString(getSd("field f1 type array<tensor(x{})> {}"));
+            fail("Expected exception");
+        }
+        catch (IllegalArgumentException e) {
+            assertEquals("For search 'test', field 'f1': A field with collection type of tensor is not supported. Use simple type 'tensor' instead.",
+                         e.getMessage());
+        }
     }
 
     @Test
     public void requireThatTensorFieldCannotBeIndexField() throws ParseException {
-        exception.expect(IllegalArgumentException.class);
-        exception.expectMessage("For search 'test', field 'f1': A field of type 'tensor' cannot be specified as an 'index' field.");
-        SearchBuilder.createFromString(getSd("field f1 type tensor(x{}) { indexing: index }"));
+        try {
+            SearchBuilder.createFromString(getSd("field f1 type tensor(x{}) { indexing: index }"));
+            fail("Expected exception");
+        }
+        catch (IllegalArgumentException e) {
+            assertEquals("For search 'test', field 'f1': A field of type 'tensor' cannot be specified as an 'index' field.",
+                         e.getMessage());
+        }
     }
 
     @Test
     public void requireThatTensorAttributeCannotBeFastSearch() throws ParseException {
-        exception.expect(IllegalArgumentException.class);
-        exception.expectMessage("For search 'test', field 'f1': An attribute of type 'tensor' cannot be 'fast-search'.");
-        SearchBuilder.createFromString(getSd("field f1 type tensor(x{}) { indexing: attribute \n attribute: fast-search }"));
+        try {
+            SearchBuilder.createFromString(getSd("field f1 type tensor(x{}) { indexing: attribute \n attribute: fast-search }"));
+            fail("Expected exception");
+        }
+        catch (IllegalArgumentException e) {
+            assertEquals("For search 'test', field 'f1': An attribute of type 'tensor' cannot be 'fast-search'.", e.getMessage());
+        }
     }
 
     @Test
     public void requireThatIllegalTensorTypeSpecThrowsException() throws ParseException {
-        exception.expect(IllegalArgumentException.class);
-        exception.expectMessage("Field type: Illegal tensor type spec: A tensor type spec must be on the form tensor[<valuetype>]?(dimensionidentifier[{}|[length?]*), but was 'tensor(invalid)'. Dimension 'invalid' is on the wrong format. Examples: tensor(x[]), tensor<float>(name{}, x[10])");
-        SearchBuilder.createFromString(getSd("field f1 type tensor(invalid) { indexing: attribute }"));
+        try {
+            SearchBuilder.createFromString(getSd("field f1 type tensor(invalid) { indexing: attribute }"));
+            fail("Expected exception");
+        }
+        catch (IllegalArgumentException e) {
+            assertStartsWith("Field type: Illegal tensor type spec:", e.getMessage());
+        }
     }
 
     private static String getSd(String field) {
         return "search test {\n document test {\n" + field + "}\n}\n";
+    }
+
+    private void assertStartsWith(String prefix, String string) {
+        assertEquals(prefix, string.substring(0, Math.min(prefix.length(), string.length())));
     }
 
 }

--- a/vespajlib/abi-spec.json
+++ b/vespajlib/abi-spec.json
@@ -1352,7 +1352,8 @@
       "public static com.yahoo.tensor.TensorType$Value[] values()",
       "public static com.yahoo.tensor.TensorType$Value valueOf(java.lang.String)",
       "public static com.yahoo.tensor.TensorType$Value largestOf(java.util.List)",
-      "public static com.yahoo.tensor.TensorType$Value largestOf(com.yahoo.tensor.TensorType$Value, com.yahoo.tensor.TensorType$Value)"
+      "public static com.yahoo.tensor.TensorType$Value largestOf(com.yahoo.tensor.TensorType$Value, com.yahoo.tensor.TensorType$Value)",
+      "public java.lang.String toString()"
     ],
     "fields": [
       "public static final enum com.yahoo.tensor.TensorType$Value DOUBLE",

--- a/vespajlib/src/main/java/com/yahoo/tensor/TensorParser.java
+++ b/vespajlib/src/main/java/com/yahoo/tensor/TensorParser.java
@@ -94,53 +94,33 @@ class TensorParser {
         if (type.get().dimensions().stream().anyMatch(d -> ( d.size().isEmpty())))
             throw new IllegalArgumentException("The dense tensor form requires a tensor type containing " +
                                                "only dense dimensions with a given size");
-        IndexedTensor.BoundBuilder builder = (IndexedTensor.BoundBuilder)IndexedTensor.Builder.of(type.get());
 
-        // Since we know the dimensions the brackets are just syntactic sugar
-        long[] indexes = new long[builder.type().rank()];
+        IndexedTensor.BoundBuilder builder = (IndexedTensor.BoundBuilder)IndexedTensor.Builder.of(type.get());
+        long index = 0;
         int currentChar;
         int nextNumberEnd = 0;
+        // Since we know the dimensions the brackets are just syntactic sugar:
         while ((currentChar = nextStartCharIndex(nextNumberEnd + 1, valueString)) < valueString.length()) {
             nextNumberEnd   = nextStopCharIndex(currentChar, valueString);
             if (currentChar == nextNumberEnd) return builder.build();
 
-            if (builder.type().valueType() == TensorType.Value.DOUBLE)
-                builder.cellByDirectIndex(nextCellIndex(indexes, builder), Double.parseDouble(valueString.substring(currentChar, nextNumberEnd)));
-            else if (builder.type().valueType() == TensorType.Value.FLOAT)
-                builder.cellByDirectIndex(nextCellIndex(indexes, builder), Float.parseFloat(valueString.substring(currentChar, nextNumberEnd)));
-            else
-                throw new IllegalArgumentException(builder.type().valueType() + " is not supported");
+            TensorType.Value cellValueType = builder.type().valueType();
+            String cellValueString = valueString.substring(currentChar, nextNumberEnd);
+            try {
+                if (cellValueType == TensorType.Value.DOUBLE)
+                    builder.cellByDirectIndex(index, Double.parseDouble(cellValueString));
+                else if (cellValueType == TensorType.Value.FLOAT)
+                    builder.cellByDirectIndex(index, Float.parseFloat(cellValueString));
+                else
+                    throw new IllegalArgumentException(cellValueType + " is not supported");
+            }
+            catch (NumberFormatException e) {
+                throw new IllegalArgumentException("At index " + index + ": '" +
+                                                   cellValueString + "' is not a valid " + cellValueType);
+            }
+            index++;
         }
         return builder.build();
-    }
-
-    // -----
-
-    /**
-     * Advance to the next cell in left-adjac ent order.
-     *
-     * On rightmost vs. leftmost adjacency:
-     * A dense tensor is laid out with the rightmost dimension as adjacent numbers,
-     * but when we parse a dense tensor we encounter numbers in the leftmost-adjacent order, since
-     * that is the most natural way to write it: tensor(x,y)[[1,2],[3,4]]
-     * should mean {{x:0, y:0}:1, {x:1, y:0}:2, {x:0, y:1}:3, {x:1, y:1}:4}.
-     * Therefore we need to convert the encounter order (numberIndex) from left-adjacent to right-adjacent.
-     */
-    private static long nextCellIndex(long[] indexes, IndexedTensor.BoundBuilder builder) {
-        long cellIndex = IndexedTensor.toValueIndex(indexes, builder.sizes());
-
-        // Find next dimension to advance
-        int nextInDimension = 0;
-        while (nextInDimension < indexes.length && indexes[nextInDimension] + 1 >= builder.sizes().size(nextInDimension)) {
-            indexes[nextInDimension] = 0;
-            nextInDimension++;
-        }
-        if (nextInDimension < indexes.length)
-            indexes[nextInDimension]++;
-        else // there is no next - become invalid
-            indexes[0]++;
-
-        return cellIndex;
     }
 
     /** Returns the position of the next character that should contain a number, or if none the string length */
@@ -187,8 +167,21 @@ class TensorParser {
             }
 
             TensorAddress address = addressBuilder.build();
-            Double value = asDouble(address, s.substring(index, valueEnd).trim());
-            builder.cell(address, value);
+            TensorType.Value cellValueType = builder.type().valueType();
+            String cellValueString = s.substring(index, valueEnd).trim();
+            try {
+                if (cellValueType == TensorType.Value.DOUBLE)
+                    builder.cell(address, Double.parseDouble(cellValueString));
+                else if (cellValueType == TensorType.Value.FLOAT)
+                    builder.cell(address, Float.parseFloat(cellValueString));
+                else
+                    throw new IllegalArgumentException(cellValueType + " is not supported");
+            }
+            catch (NumberFormatException e) {
+                throw new IllegalArgumentException("At " + address.toString(builder.type()) + ": '" +
+                                                   cellValueString + "' is not a valid " + cellValueType);
+            }
+
             index = valueEnd+1;
             index = skipSpace(index, s);
         }
@@ -217,15 +210,6 @@ class TensorParser {
                                                    "got '" + elementString + "'");
             String dimension = pair[0].trim();
             builder.add(dimension, pair[1].trim());
-        }
-    }
-
-    private static Double asDouble(TensorAddress address, String s) {
-        try {
-            return Double.valueOf(s);
-        }
-        catch (NumberFormatException e) {
-            throw new IllegalArgumentException("At " + address + ": Expected a floating point number, got '" + s + "'");
         }
     }
 

--- a/vespajlib/src/main/java/com/yahoo/tensor/TensorType.java
+++ b/vespajlib/src/main/java/com/yahoo/tensor/TensorType.java
@@ -48,6 +48,9 @@ public class TensorType {
             return FLOAT;
         }
 
+        @Override
+        public String toString() { return name().toLowerCase(); }
+
     };
 
     /** The empty tensor type - which is the same as a double */

--- a/vespajlib/src/test/java/com/yahoo/tensor/TensorParserTestCase.java
+++ b/vespajlib/src/test/java/com/yahoo/tensor/TensorParserTestCase.java
@@ -31,36 +31,36 @@ public class TensorParserTestCase {
                      Tensor.from("tensor(x[2]):[1.0, 2.0]"));
         assertEquals(Tensor.Builder.of(TensorType.fromSpec("tensor(x[2],y[3])"))
                                    .cell(1.0, 0, 0)
-                                   .cell(2.0, 1, 0)
-                                   .cell(3.0, 0, 1)
-                                   .cell(4.0, 1, 1)
-                                   .cell(5.0, 0, 2)
+                                   .cell(2.0, 0, 1)
+                                   .cell(3.0, 0, 2)
+                                   .cell(4.0, 1, 0)
+                                   .cell(5.0, 1, 1)
                                    .cell(6.0, 1, 2).build(),
                      Tensor.from("tensor(x[2],y[3]):[[1.0, 2.0], [3.0, 4.0], [5.0, 6.0]]"));
         assertEquals(Tensor.Builder.of(TensorType.fromSpec("tensor(x[1],y[2],z[3])"))
                                    .cell(1.0, 0, 0, 0)
-                                   .cell(2.0, 0, 1, 0)
-                                   .cell(3.0, 0, 0, 1)
-                                   .cell(4.0, 0, 1, 1)
-                                   .cell(5.0, 0, 0, 2)
+                                   .cell(2.0, 0, 0, 1)
+                                   .cell(3.0, 0, 0, 2)
+                                   .cell(4.0, 0, 1, 0)
+                                   .cell(5.0, 0, 1, 1)
                                    .cell(6.0, 0, 1, 2).build(),
                      Tensor.from("tensor(x[1],y[2],z[3]):[[[1.0], [2.0]], [[3.0], [4.0]], [[5.0], [6.0]]]"));
         assertEquals(Tensor.Builder.of(TensorType.fromSpec("tensor(x[3],y[2],z[1])"))
                                    .cell(1.0, 0, 0, 0)
-                                   .cell(2.0, 1, 0, 0)
-                                   .cell(3.0, 2, 0, 0)
-                                   .cell(4.0, 0, 1, 0)
-                                   .cell(5.0, 1, 1, 0)
+                                   .cell(2.0, 0, 1, 0)
+                                   .cell(3.0, 1, 0, 0)
+                                   .cell(4.0, 1, 1, 0)
+                                   .cell(5.0, 2, 0, 0)
                                    .cell(6.0, 2, 1, 0).build(),
                      Tensor.from("tensor(x[3],y[2],z[1]):[[[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]]"));
         assertEquals(Tensor.Builder.of(TensorType.fromSpec("tensor(x[3],y[2],z[1])"))
-                                   .cell(1.0, 0, 0, 0)
-                                   .cell(2.0, 1, 0, 0)
-                                   .cell(3.0, 2, 0, 0)
-                                   .cell(4.0, 0, 1, 0)
-                                   .cell(5.0, 1, 1, 0)
-                                   .cell(6.0, 2, 1, 0).build(),
-                     Tensor.from("tensor( x[3],y[2],z[1]) : [  [ [1.0, 2.0, 3.0] , [4.0, 5,6.0] ]  ]"));
+                                   .cell( 1.0, 0, 0, 0)
+                                   .cell( 2.0, 0, 1, 0)
+                                   .cell( 3.0, 1, 0, 0)
+                                   .cell( 4.0, 1, 1, 0)
+                                   .cell( 5.0, 2, 0, 0)
+                                   .cell(-6.0, 2, 1, 0).build(),
+                     Tensor.from("tensor( x[3],y[2],z[1]) : [  [ [1.0, 2.0, 3.0] , [4.0, 5,-6.0] ]  ]"));
     }
 
     @Test
@@ -71,6 +71,10 @@ public class TensorParserTestCase {
                       "{{'x':\"l0\"}:1.0}");
         assertIllegal("dimension must be an identifier or integer, not '\"x\"'",
                       "{{\"x\":\"l0\", \"y\":\"l0\"}:1.0, {\"x\":\"l0\", \"y\":\"l1\"}:2.0}");
+        assertIllegal("At {x:0}: '1-.0' is not a valid double",
+                      "{{x:0}:1-.0}");
+        assertIllegal("At index 0: '1-.0' is not a valid double",
+                      "tensor(x[1]):[1-.0]");
     }
 
     private void assertIllegal(String message, String tensor) {

